### ISSUE TITLE
Support federated clusters in separate namespaces

### ISSUE THREAD

**Issue by andrewmyhre** [kubernetes-sigs/kubefed#526]
In our use case we want to run 'fleets' of kubernetes clusters which would be owned by different groups which only have permissions to get / list / update / delete within their fleet. By creating federated cluster resources in separate namespaces in the federation cluster we could leverage RBAC to scope user access to different clusters. 

Currently the federation controller is scoped to one namespace, so we would need to run a separate controller in each federation namespace. That's fine with me but it would be great if it could act as a controller for federated clusters in multiple namespaces (or just all namespaces). Also, currently if you try to deploy the controller helm chart to two different namespaces you get a conflict error:

```
customresourcedefinitions.apiextensions.k8s.io "federatedclusterroles.primitives.federation.k8s.io" already exists
```

Wondering if this is a use case which could be in scope. Happy to contribute a PR if so.

**Comment by font** [comment]
The federation controller by default supports running as one instance to monitor all namespaces. We tend to refer to this as a cluster-scoped deployment. The namespace-scoped deployment that you describe is the other supported mode.

Ah yes, that error is because custom resource definitions are cluster scoped. For namespace-scoped deployments i.e. with `limitedScope=true`, we should not fail if the CRDs are already installed. Additionally, we should probably not even install CRDs for cluster-scoped target resources such as the one from your error. @gyliu513 for the helm chart.

**Comment by andrewmyhre** [comment]
Aha, so if I'm understanding correctly this model would work for me as-is if i dropped the `--limited-scope` argument. I'll try this. Thanks!

**Comment by andrewmyhre** [comment]
@font so I actually had to make a couple of changes to get this to work, which could be because I'm joining clusters incorrectly. The namespace where federated clusters are listed, watched and where statuses are updated are locked to the federated cluster namespace, which means if you add a federated cluster to any other namespace the controller will ignore them. The cluster secrets have the same problem.

Here's a commit with the changes I made: https://github.com/andrewmyhre/federation-v2/commit/13202ed7adde8fc72fb03780116ad44edc6e8ac6

I'm probably not understanding how the `limited-scope=false` mode is supposed to work but it seems to me these changes are required for that mode to work. Interested to know what you think.

**Comment by font** [comment]
@andrewmyhre Are you using `kubefed2 join` to join your clusters to the federation? If so, the `--federation-namespace` defaults to `federation-system` and currently needs to match the namespace that the federation controller manager is running in.

Is it important for you to use different namespaces to create federated clusters? Or are you able to use the same namespace as the controller manager?

**Comment by marun** [comment]
@andrewmyhre Sourcing the clusters from a single namespaces ensures that a given cluster name can exist at most once for a given federation control plane.  Ensuring cluster name uniqueness is required to support placement via `clusterName`, as name collisions would result in unexpected behavior.

Your proposed change would also break namespaced (`--limited-scope=true`) deployment of the federation control plane, since in that mode the control plane is not granted permission to more than the namespace in which it is deployed.

`--limited-scope=true` (namespaced) deployment implies deploying the fedv2 control plane in a single namespace with access to only that namespace and targeting only that namespace and its contents for propagation to member clusters.  `--limited-scope=false` (cluster-wide) deployment implies deploying the control plane in a single namespace but with access to all resources in the cluster and the ability to propagate any namespace and its contents to member clusters. 

If I'm understanding correctly, it sounds you would like to define associativity between a cluster and one or more namespaces, and only permit propagation to that cluster from those namespaces.  While there is no mechanism at present to define associativity at the cluster level, it could be definable for each namespace.  The `FederatedNamespacePlacement` resource for each namespace declares the clusters a namespace should be propagated to (either by `clusterNames` or `clusterSelector`), and contained resources can only be propagated to a target cluster if the namespace exists on that cluster.  It is currently possible that a resource be propagated to a target cluster even if the namespace placement doesn't include that cluster so long as the target cluster has the namespace already created, but that could be fixed by computing placement as the union of namespace and resource placement.  

Provided placement computed as the union of namespace and resource placement, and RBAC protection of namespace placement to prevent user modification, propagation of the contents of namespaces could be limited to a subset of available clusters.  Would that satisfy your use case?

**Comment by gyliu513** [comment]
@andrewmyhre would you please help check if the document here https://docs.google.com/document/d/1s_w3auALOt7U0Ifu6oCh4-NcF0JwDU8-SAAhtxOUQKc/edit?usp=sharing can satisfy your requirement?

**Comment by andrewmyhre** [comment]
@marun 

> If I'm understanding correctly, it sounds you would like to define associativity between a cluster and one or more namespaces, and only permit propagation to that cluster from those namespaces.

That's partly it, yes. I'll provide some background..

I work in a large organization (financial, process driven, tight access controls etc) where Kubernetes is quickly becoming a common capability. Currently there are hundreds of AWS accounts scoped for various lines of business and areas within those LOBs. Hundreds of teams are looking to start using Kubernetes clusters and they're springing up all over the organization. My team's mission is to provide 'platform' capabilities for those clusters across the organization. We don't have any control over teams just spinning up their own clusters to run a few applications, though.

So let's say there are 1000 Kubernetes clusters and, say, 300 cluster operators. I want to provide them a dashboard where they log in with their SSO credentials and they see a list of their clusters, they can edit resources etc. I want the list of clusters they see to only be the 3 or so clusters they actually operate, not all 1000. It seemed to me that grouping the FederatedCluster resources within namespaces and then applying RoleBindings within those namespaces to manage access control would be a good Kubernetes-y way to accomplish that.

TL;DR: grouping clusters by namespace would provide RBAC on clusters which we would need to use this in our org.

Maybe there's already a way to create a rolebinding to provide access to a specific set of resources (as opposed to resource _types_) with a single namespace that I'm not aware of? That would solve our problem. My understanding was role bindings are scoped to resource types and namespaces.

Totally understand now that it would break placement. I'm wondering if supporting a `<namespace>/<cluster name>` value in the placement cluster names list is an option.

Yes it would also break `--limited-scope=true` *if* someone used that setting and then joined clusters in different namespaces. To me that would be a case of user error, no?

**Comment by marun** [comment]
@andrewmyhre It is possible to control access to a specific set of resources via RBAC:

https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources

It should be possible, then, to create an RBAC role defining the subset of federated clusters (defined centrally in the federation system namespace) that a given operator should have access to. In combination with the previously described namespace placement change, it should be possible to both ensure authz to a subset of member clusters and restrict propagation to that same subset.  Would this support your use case in the absence of being able to define federated clusters within arbitrary namespaces?  

I'm not sure I understand a clean way to associate clusters to namespaces otherwise, since putting a cluster in an arbitrary namespace doesn't provide an indication of what other namespaces should be permitted to be propagated to that cluster.

**Comment by andrewmyhre** [comment]
Finally caught up on this. I had the wrong understanding of namespace scoped deployments when I first raised this issue. Thanks for taking the time to explain!
I'll read the document @gyliu513 shared and also look at a way to enforce RBAC on kubefedclusters as they're created/updated/deleted (custom controller I'm thinking).
